File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_set<char> charSet;
        
        int l = 0;
        
        int res = 0;
        
        for(int r=0; r<s.length();r++){
            while(charSet.find(s[r]) != charSet.end()){
                charSet.erase(s[l]);
                l++;
            }
            
            charSet.insert(s[r]);
            res = max(res , r-l+1);
        }
        return res;
        
    }
};

/*
int main() {
    Solution solution;

    string input_str = "abcabcbb";

    int result = solution.lengthOfLongestSubstring(input_str);

    cout << "Length of the longest substring without repeating characters: " << result << endl;

    return 0;
}
*/

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
## time:o(n)
## space: o(n)
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Create a hashset to avoid duplicates
        charSet = set()
        
        # Initialize the result variable
        res = 0
        
        # Initialize the left pointer
        l = 0
        
        # Iterate through every character (right pointer)
        for r in range(len(s)):
            # Check if the character is already in the set
            while s[r] in charSet:
                # Remove the leftmost character from the set
                charSet.remove(s[l])
                # Increment the left pointer
                l += 1
                
            # Add the current character to the set
            charSet.add(s[r])
            
            # Update the result with the maximum length
            res = max(res, r - l + 1)
                
        return res
